resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopGlobalFeatureDriverStep.scala

```diff
@@ -0,0 +1,151 @@
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.k8s.features
+
+import java.io.File
+
+import scala.collection.JavaConverters._
+
+import com.google.common.base.Charsets
+import com.google.common.io.Files
+import io.fabric8.kubernetes.api.model.{ConfigMapBuilder, HasMetadata}
+
+import org.apache.spark.deploy.k8s.{KubernetesConf, KubernetesUtils, SparkPod}
+import org.apache.spark.deploy.k8s.Config._
+import org.apache.spark.deploy.k8s.Constants._
+import org.apache.spark.deploy.k8s.KubernetesDriverSpecificConf
+import org.apache.spark.deploy.k8s.features.hadoopsteps._
+import org.apache.spark.internal.Logging
+
+ /**
+  * Runs the necessary Hadoop-based logic based on Kerberos configs and the presence of the
+  * HADOOP_CONF_DIR. This runs various bootstrap methods defined in HadoopBootstrapUtil.
+  */
+private[spark] class HadoopGlobalFeatureDriverStep(
+    kubernetesConf: KubernetesConf[KubernetesDriverSpecificConf])
+    extends KubernetesFeatureConfigStep with Logging {
+
+    private val conf = kubernetesConf.sparkConf
+    private val maybePrincipal = conf.get(org.apache.spark.internal.config.PRINCIPAL)
+    private val maybeKeytab = conf.get(org.apache.spark.internal.config.KEYTAB)
+    private val maybeExistingSecretName = conf.get(KUBERNETES_KERBEROS_DT_SECRET_NAME)
+    private val maybeExistingSecretItemKey =
+      conf.get(KUBERNETES_KERBEROS_DT_SECRET_ITEM_KEY)
+    private val kubeTokenManager = kubernetesConf.tokenManager
+    private val isKerberosEnabled = kubeTokenManager.isSecurityEnabled
+
+    require(maybeKeytab.forall( _ => isKerberosEnabled ),
+      "You must enable Kerberos support if you are specifying a Kerberos Keytab")
+
+    require(maybeExistingSecretName.forall( _ => isKerberosEnabled ),
+      "You must enable Kerberos support if you are specifying a Kerberos Secret")
+
+    KubernetesUtils.requireBothOrNeitherDefined(
+      maybeKeytab,
+      maybePrincipal,
+      "If a Kerberos principal is specified you must also specify a Kerberos keytab",
+      "If a Kerberos keytab is specified you must also specify a Kerberos principal")
+
+    KubernetesUtils.requireBothOrNeitherDefined(
+      maybeExistingSecretName,
+      maybeExistingSecretItemKey,
+      "If a secret data item-key where the data of the Kerberos Delegation Token is specified" +
+        " you must also specify the name of the secret",
+      "If a secret storing a Kerberos Delegation Token is specified you must also" +
+        " specify the item-key where the data is stored")
+
+    require(kubernetesConf.hadoopConfDir.isDefined, "Ensure that HADOOP_CONF_DIR is defined")
+    private val hadoopConfDir = kubernetesConf.hadoopConfDir.get
+    private val hadoopConfigurationFiles = kubeTokenManager.getHadoopConfFiles(hadoopConfDir)
+
+    // Either use pre-existing secret or login to create new Secret with DT stored within
+    private val hadoopSpec: Option[KerberosConfigSpec] = (for {
```

Comment:
Can bundle secret name and secret key in a case class so that for... yield isn't necessary.

Reply:
In this specific case, the `for..yield` is quite clear IMHO. I think it is easier to parse. 
I would prefer to leave it.